File: 006-binary-search-tree/binary-search-tree.go
package main

import "fmt"

// Node represents node of tree
type Node struct {
	key    int
	left   *Node
	right  *Node
	parent *Node
}

// BinarySearchTree represents Tree
type BinarySearchTree struct {
	root *Node
}

// NewNode return reference to new node
func (t *BinarySearchTree) NewNode(key int) *Node {
	return &Node{
		key: key,
	}
}

// InorderTraverse tree
func (t *BinarySearchTree) InorderTraverse(node *Node) {
	if node != nil {
		t.InorderTraverse(node.left)
		fmt.Printf("%d\t", node.key)
		t.InorderTraverse(node.right)
	}
}

// Search for node
func (t *BinarySearchTree) Search(key int) *Node {
	ptr := t.root
	for ptr != nil && ptr.key != key {
		if key < ptr.key {
			ptr = ptr.left
		} else {
			ptr = ptr.right
		}
	}
	return ptr
}

// FindMin key
func (t *BinarySearchTree) FindMin(ptr *Node) *Node {
	for ptr != nil && ptr.left != nil {
		ptr = ptr.left
	}
	return ptr
}

// FindMax key
func (t *BinarySearchTree) FindMax(ptr *Node) *Node {
	for ptr != nil && ptr.right != nil {
		ptr = ptr.right
	}
	return ptr
}

// FindInorderSuccessor of given node
func (t *BinarySearchTree) FindInorderSuccessor(ptr *Node) *Node {
	if ptr != nil {
		if ptr.right != nil {
			return t.FindMin(ptr.right)
		}
		y := ptr
		x := ptr.parent
		for x != nil && x.right == y {
			y = x
			x = x.parent
		}
		return x
	}
	return ptr
}

// FindInorderPredecessor of given node
func (t *BinarySearchTree) FindInorderPredecessor(ptr *Node) *Node {
	if ptr != nil {
		if ptr.left != nil {
			return t.FindMax(ptr.left)
		}
		y := ptr
		x := ptr.parent
		for x != nil && x.left == y {
			y = x
			x = x.parent
		}
		return x
	}
	return ptr
}

// Insert key in tree
func (t *BinarySearchTree) Insert(key int) {
	ptr := t.root
	var parent *Node = nil
	for ptr != nil {
		parent = ptr
		if key < ptr.key {
			ptr = ptr.left
		} else {
			ptr = ptr.right
		}
	}
	newNode := t.NewNode(key)
	if parent == nil {
		t.root = newNode
	} else if key < parent.key {
		parent.left = newNode
	} else {
		parent.right = newNode
	}
	newNode.parent = parent
}

// Transpalant changes parent-child relationship
func (t *BinarySearchTree) Transpalant(u *Node, v *Node) {
	if u.parent == nil {
		t.root = v
	} else if u.parent.left == u {
		u.parent.left = v
	} else {
		u.parent.right = v
	}

	if v != nil {
		v.parent = u.parent
	}
}

// Delete key in tree
func (t *BinarySearchTree) Delete(key int) (int, error) {
	node := t.Search(key)
	if node == nil {
		return key, fmt.Errorf("key not found: %d", key)
	}
	if node.left == nil {
		t.Transpalant(node, node.right)
	} else if node.right == nil {
		t.Transpalant(node, node.left)
	} else {
		successor := t.FindInorderSuccessor(node)
		if node.right != successor {
			t.Transpalant(successor, successor.right)
			successor.right = node.right
			successor.right.parent = successor
		}
		t.Transpalant(node, successor)
		successor.left = node.left
		node.left.parent = successor
	}
	return key, nil
}

// NewTree returns reference to new BST
func NewTree() *BinarySearchTree {
	return &BinarySearchTree{}
}

func main() {

	bst := NewTree()

	fmt.Println("Tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Tree deletion 6:: ")
	_, err := bst.Delete(6)
	if err != nil {
		fmt.Println(err)
	}

	fmt.Println("Inserting 40")
	bst.Insert(40)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 20")
	bst.Insert(20)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 70")
	bst.Insert(70)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 80")
	bst.Insert(80)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 50")
	bst.Insert(50)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 45")
	bst.Insert(45)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 90")
	bst.Insert(90)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 75")
	bst.Insert(75)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 72")
	bst.Insert(72)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Inserting 74")
	bst.Insert(74)
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Tree deletion 45 :: ")
	_, err = bst.Delete(45)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Tree deletion 9 :: ")
	_, err = bst.Delete(9)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Tree deletion 70 :: ")
	_, err = bst.Delete(70)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Tree deletion 72 :: ")
	_, err = bst.Delete(72)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Tree deletion 74 :: ")
	_, err = bst.Delete(74)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")

	fmt.Println("Tree deletion 80 :: ")
	_, err = bst.Delete(80)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println("tree traversal :: ")
	bst.InorderTraverse(bst.root)
	fmt.Println("")
}

File: 006-binary-search-tree/binary-search-tree.js
class Node {
	key;
	left;
	right;

	constructor (key) {
		this.key = key;
		this.left = this.right = null;
	}
}

class BST {
	root;
	traverseData = [];

	constructor() {
		this.root = null;
	}

	insert(key) {
		const newNode = new Node(key);
		let x = this.root;
		let y = null;
		while (x !== null) {
			y = x;
			if (key <= x.key) {
				x = x.left;
			} else {
				x = x.right;
			}
		}
		if (y === null) {
			this.root = newNode;
		} else if (key <= y.key) {
			y.left = newNode;
		} else {
			y.right = newNode;
		}
	}

	inorderTraverse(node) {
		if (node !== null) {
			this.inorderTraverse(node.left);
			//console.log(node.key);
			this.traverseData.push(node.key);
			this.inorderTraverse(node.right);
		}
	}

	search(key) {
		let node = this.root;
		let parent = null;
		while (node !== null && node.key !== key) {
			parent = node;
			if (key < node.key) {
				node = node.left;
			} else {
				node = node.right;
			}
		}
		return {
			node,
			parent
		};
	}

	delete(key) {
		const  { node, parent } = this.search(key);
		if (node !== null) {
			if (node.left === null) {
				this.transplant(parent, node, node.right);
			} else if (node.right === null) {
				this.transplant(parent, node, node.left);
			} else {
				let inorderSuccParent = node;
				let inorderSucc = node.right;
				while (inorderSucc.left !== null) {
					inorderSuccParent = inorderSucc;
					inorderSucc = inorderSucc.left;
				}
				if (inorderSuccParent !== node) {
					this.transplant(inorderSuccParent, inorderSucc, inorderSucc.right);
					inorderSucc.right = node.right;
				}
				inorderSucc.left = node.left;
				this.transplant(parent, node, inorderSucc);
			}
		}
	}

	transplant(parent, u, v ) {
		if (parent === null) {
			this.root = v;
		} else if (u.key <= parent.key) {
			parent.left = v;
		} else {
			parent.right = v;
		}
	}
}

const tree1 = new BST();

tree1.insert(80);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(20);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(240);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(180);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(325);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(310);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(300);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(312);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(314);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(305);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(302);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(308);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(40);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(16);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(14);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(30);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(80);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.insert(75);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.delete(308);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log('After delete 308');
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.delete(16);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log('After delete 16');
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.delete(312);
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log('After delete 312');
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.delete(240); // left and right child both are not null
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log('After delete 240');
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.delete(310); // left and right child both are not null and right child is inorder successor
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log('After delete 310');
console.log(tree1.traverseData.join(', '));
console.log('=======');

tree1.delete(80); // delete root
tree1.traverseData = [];
tree1.inorderTraverse(tree1.root);
console.log('After delete 80');
console.log(tree1.traverseData.join(', '));
console.log('=======');

console.log('search-314', tree1.search(314));
console.log('=======');

console.log('search-79', tree1.search(79));
console.log('=======');

console.log('search-75', tree1.search(75));
console.log('=======');